### ISSUE TITLE
Request for documentation, manuals, and original media distribution.

### ISSUE THREAD

**Issue by the-grue** [roelandjansen/pcmos386v501#27]
I believe it would be very helpful when trying to recreate a distribution of the software to be able to have the full layout from an actual full distribution.  I would be willing to work on distribution tools and scripts to generate media if someone could locate and provide access to a full set of original disk images.

We could also really use original documentation or manuals scanned to PDF.  Vendor build documentation, tech notes, etc. would be awesome, but I would imagine they are lost to the ages.

Another boon would be scans of any books related to the OS, if they exist.

**Comment by andrewbird** [comment]
Winworld seems to have images of various PC-MOS disks. I've not tried any of them myself, but may give you an indication of the layout https://winworldpc.com/product/pc-mos/5

**Comment by andrewbird** [comment]
There's also a little mention of PC-MOS setup in this book https://ia801901.us.archive.org/26/items/DOS_Beyond_640K_2nd_edition/DOS_Beyond_640K_2nd_edition.pdf at page 200-203

**Comment by the-grue** [comment]
Thanks Andrew!  I came across both of those in my searching online.  Can anyone confirm if the disk on the Winworld site is legit and complete?

**Comment by the-grue** [comment]
Came across a posting on the Vintage Computer Forum where someone with the handle sharkonwheels was in a discussion with Roeland about this software many years ago.  This person also stated that they had original manuals and software.  I tried to reach out to them through the Vintage Computer Forum, but had no luck.  It doesn't look like they have logged into that forum since January of this year.  If anyone has contact information for this person, it might be worth it to reach out to them and ask if they could provide scans of the documents or would be willing to loan them out to someone to scan.

The search continues...

**Comment by roelandjansen** [comment]
any news on it?

**Comment by the-grue** [comment]
No, unfortunately sharkonwheels hasn't logged into the site since January 15, 2018.  You wouldn't happen to have contact information outside that forum, would you?

**Comment by the-grue** [comment]
Happy New Year!  Still nothing.  Just posted to the Vintage Computer forum to see if anyone on there can help.

**Comment by the-grue** [comment]
Saw a copy of PC-MOS/386 v4 on ebay for $49.  Not sure how close v4 is to v5, so I didn't take a chance on it.

**Comment by the-grue** [comment]
Does anyone know if the v4 manual(s) would be close enough to v5 to merit buying this?  If so, I will try to make the purchase and scan/upload to the project.

**Comment by src153** [comment]
> There's also a little mention of PC-MOS setup in this book https://ia801901.us.archive.org/26/items/DOS_Beyond_640K_2nd_edition/DOS_Beyond_640K_2nd_edition.pdf at page 200-203

It says some bugs were gone by the 4.1 release. It also says you need the manual to know what you are doing. That Ebay sale sounds pretty good ...

**Comment by roelandjansen** [comment]
It's also interesting to note that 

http://www.cs.cmu.edu/~ralf/files.html

contains the files to the interrupts list of the API et al. In the past I have managd to get it in the book as well (printed as well as PDF/html)

Maybe I should add it to the repo as well. Just for the fun of it. Same holds for the above mentioned book.

**Comment by src153** [comment]
Roeland, do know the important changes from 4.x to 5.x?

Also, I would suggest removing the BORLANDC directory altogether and just noting on the project page that it's available at Winworld. I posted a link in #30. Borland make was the first stumbling point for me when building the kernel for the first time. That mostly empty directory suggested to me that I had to use my own Borland directory, which is misleading to a newcomer trying to build the kernel.

**Comment by roelandjansen** [comment]
I seriously don't know anymore.  I started working on v5.x at that time.

When it comes to borland, other tooling -- some time ago I started checking if I could copyright-wise could add it to the repo so that the whole was in a fully working/buildable state.

It even would include brief as the editor. 

Now, I have been ill for over a year with a few operations so it became a background task I guess.

**Comment by the-grue** [comment]
I ordered the v4 software/manual.  Should be here next week.

Funny you should mention the interrupt list.  I was looking at the MOS API services in the code last night and on the interrupt list while trying to track down the source of the 2012 bug.  I didn't know MOS had these features and had been focusing on int 21.

Here is where API and data structure documentation would come in handy.  It looks like MOS stores the user information and class in the reserved section of the FAT directory entry, but I am not 100% sure yet.  That has been where I have been focused and I still believe that is where I will root out the bug.

I need more information on how the class entry works and if it is encrypting data blocks.

**Comment by src153** [comment]
If there is no second project owner, I nominate the-grue. His picture looks young and healthy. I'm too old, probably older than Roeland.

I don't know how these things work at github, I'm new to it. I did create a hello-world project when following their tutorial, and then I deleted it. All it took was a confirmation click, and the whole project was gone forever. That's why I saved a .ZIP of this one.

**Comment by spacerace** [comment]
About Borland Tools:
If I remember correctly the agreement when registering on github, it is not allowed to host shareware/propritary software etc where you don't own the rights.
As other mentioned, I would just write a short text file on how to setup Borland's tools and refer to winworld.
Otherwise the project would be in danger of being deleted. Maybe this won't happen, but there's always a small chance....

@roelandjansen Hope it will get better!

**Comment by roelandjansen** [comment]
I just recalled that a good old friend of mine, Arjen Schouten, might have the manuals and orignal disks so that it's exactly known what was on the disks originally. 

He'll see tomorrow if he indeed has them. Afaik it's v5.01.

**Comment by roelandjansen** [comment]
@spacerace -- I just didn't want to go there -- maybe  a legal link would be OK indeed.

**Comment by the-grue** [comment]
Probably better to stick with what we have and not put the project in jeopardy by adding anything (else) that may be considered IP.  Folks can easily find the tools needed to build parts of the project and the bulk of it builds with the tools provided in the bin directory.

@roelandjansen that would be great if he can come up with anything!  Fingers crossed!

**Comment by roelandjansen** [comment]
he found it and will send it over to me. I have a printer/scanner with feeder so think that it can be done to have the quick-start and user manual scanned.

It's for v5.01 by the way

**Comment by src153** [comment]
Sounds good. But I still want to see the 4.x docs and disks. @the-grue if the 5.01 docs discourage you from scanning your 4.x docs, I will buy the package from you. At a discount of course.

**Comment by roelandjansen** [comment]
Cool, that makes the eventual diff possible too. Regarding scanning -- it probably is quite some work and I would need to look into how to OCR from the scanned PDF's I am going to make. 

This is the moment I hate myself that I ditched the books long time ago ;-)

**Comment by the-grue** [comment]
v4 stuff showed up in the mail today!  Took a bit of data recovery on the v4.1 system disk as it was throwing read errors right and left, but I think I managed to get all the bits in the end.  Aux disk copied fine first try.  Surprisingly, the smaller envelope is still sealed!  I was able to get it to install and verify that the 2012 bug appears to have been there in v4.1.  

![pcmosv4 1](https://user-images.githubusercontent.com/3101080/51578742-68cf5000-1e73-11e9-97bb-4c937ac58ee2.png)

The 5.25 disks appear to be v4.0.  I will have to resurrect my old Compaq Portable to read those at some point.  The 5.25 envelope was opened and contained a 5.25 disk.

Will work on starting the scans this weekend!

**Comment by src153** [comment]
That's good news. How did you work around the floppy read errors?

**Comment by the-grue** [comment]
Persistence, needle nosed pliers, and some compressed air applied from a safe distance.

**Comment by src153** [comment]
I understand persistence and canned air. The pliers have me wondering though. Maybe a different floppy drive would help. I have several PCs with floppies, if you're not 100% sure about getting all the bits. I could read it and we could binary compare.

**Comment by the-grue** [comment]
Pliers to turn the disk in its sleeve using the keyholes on the bottom.  I am pretty sure I got it all though, because it installed completely after I got a good dd.  I have other floppy drives I can try too, but thanks for the offer.

I am guessing this stuff sat for awhile as there is a good accumulation of dust and stuff in the manual binder where the plastic disk sleeve is.

**Comment by src153** [comment]
Holes to turn the disk, right. That hit me after I thought more about it. Using DD means linux and it's know for its poor floppy driver. There are not that many files on the system disk, are there? You could DOS copy the files with a wildcard *.* and compare against the files on your image. I'm not saying you should, I'm just saying I would, for my own peace of mind.

**Comment by the-grue** [comment]
Just uploaded the disk images for v4 (#45 ) and the documentation (#46 ).

They are available in my fork on the v4images and documentation branches respectively.

**Comment by src153** [comment]
I see Roeland merged them. They're good to have. I can't see the forest for the trees without them. Github is weird with a PDF though. It (very slowly) displays pages in the browser, when all I want to do is download and view the whole file with Adobe. The download link works, but github doesn't know that's all you want, and tries to be "helpful".

**Comment by src153** [comment]
I'm reading the user manual now. It's worth it's weight in gold. It gives you the big picture and many details. No one should submit patches without reading it.

**Comment by src153** [comment]
I read the first few chapters, skimmed the parts I don't need just yet, and I'm into chapter 9, security. The .class command acts on directories or partitions, but to set the class of a file, you must use the .copy command. I thought that seemed weird, until I discovered there is a form of the .copy command that does not a copy file:

.copy z.txt /c-

sets the class to blank. That's all it does. No duplicate file created. The /c parameter can also set the class to a non-blank value. It's all there in the manual.

**Comment by src153** [comment]
We have the 4.1 user manual now. Good enough to get started. Closing.

**Comment by the-grue** [comment]
Hi Roeland!  Any luck scanning the v5 media or documentation?

If you want, I can do the work, you'll just need to get the media to me temporarily.

**Comment by the-grue** [comment]
I just scored a set of 5.01 UPD6 media and documentation from ebay for right around the price I paid for the v4 stuff!  Should be here next week sometime.  I'll scan and upload when I get a chance.

**Comment by the-grue** [comment]
Got the documentation and media!  Unfortunately, my scanner crapped out, so it will take some time to get the scanning done using alternate methods.  In the meantime, if anyone has specific questions, feel free to ask.

**Comment by roelandjansen** [comment]
super! I found out myself that I cannot scan two ways and  it got moved to the future for this moment.

**Comment by the-grue** [comment]
Uploaded scans of the Booster documentation and install guide for v5.01.  See #79

**Comment by Sharkonwheels** [comment]
Sorry, gents. Life took over, and I don't get to check into VCF much these days. I see you've gotten the v5 docs. FYI, the 5.01 PC-MOS floating around the InterWebz is probably my copy. It spread like wildfire. It's a 9-user version, and if you want to activate it to and the 30min demo stuff, once installed do "init N9AM" and it will turn it into a real copy. I do have a few IMA files for you, from the Guide to Operations manual's accompanying disk, a disk image of EmuLink, the DOS PC-MOS client, and I also have a special gift: PC-MOS Troubleshooting Guide in PDF format. This was a relatively tightly-held doc inside TSL, and as we were a dealer in Miami I convinced them i needed a copy. This assists in troubleshooting the high-memory installs and getting the really high amounts of available memory for tasks. However, I'm not versed in Git or Github, so I'll need some handholding.

**Comment by the-grue** [comment]
Welcome to the project @Sharkonwheels!  Good to see you are still around!  If you want, you can e-mail what you have to me and I will create a pull request for @roelandjansen to add to the repo or you can try to create your own from a fork->branch->commit->pull request.

Either way, I am sure we can get a lot out of the materials you have to offer.

**Comment by Sharkonwheels** [comment]
Cool. Email me using my username here, at gmail, and I'll reply back with what I have found.
One thing I have that I have NOT seen in the wild, is the PC-MOS Gateway for Novell. It uses the old-stle wsobj/wsgen, and no odi, so a client has to be "generated." What it's actually doing is linking the OBJ files with a network driver.

**Comment by roelandjansen** [comment]
the docs are more than welcome!

I had it in the past as a hardcopy but has been lost ..

**Comment by the-grue** [comment]
I got the files and have added them in pull request #80

**Comment by roelandjansen** [comment]
merged

**Comment by the-grue** [comment]
Thanks @roelandjansen!  As time permits, I will work on scanning the v5 User Guide and get it uploaded.  Then we can probably put this issue to rest.

**Comment by the-grue** [comment]
Sorry it took so long to get this done.  I have finally gotten the v5 User Guide scanned and OCR'd and uploaded.  See pull request #81

**Comment by the-grue** [comment]
I see this was merged.  I am going to close this issue.

**Comment by RenownMDK** [comment]
> I believe it would be very helpful when trying to recreate a distribution of the software to be able to have the full layout from an actual full distribution. I would be willing to work on distribution tools and scripts to generate media if someone could locate and provide access to a full set of original disk images.
> 
> We could also really use original documentation or manuals scanned to PDF. Vendor build documentation, tech notes, etc. would be awesome, but I would imagine they are lost to the ages.
> 
> Another boon would be scans of any books related to the OS, if they exist.

**Comment by RenownMDK** [comment]
I probably have the latest version of PC-Mos on floppies as well as manuals..Just came across this forum by chance

**Comment by the-grue** [comment]
Welcome!  Check what you have, what we have, and feel free to contribute anything you see that is missing.  Feel free to contact me via e-mail if you want to chat.

**Comment by RenownMDK** [comment]
I am not a programmer or even a computer guy so GirHub is mostly greek to me.
I wasn’t  sure how to contact anyone directly so I just e-mailed Roeland Jansen directly..
“Stumbled across your work on Updating PC MOS

I have been using pc-mos since early 90’s for Multi-user Inventory / accounting system (SBT on Foxpro 2.6)
We had the fatal date issue August 1 2012, and managed to carry on by setting the CMOS clock back 12 years, then adding
x weeks to the computer date in the Foxpro boot up file. This worked fine as the file date is not really important, but obviously 
we need the program dates to be current.  All going well until we hit August 2 (?) 2007 on the computer clock..same problem
all over.  We always thought this was some kind of an end-of-life date built into PC MOS, but now makes sense that it is a bug 
in the code. Fix will be harder this time!  I Would love to get the patch I read about somewhere in GitHub by “the-grue” and “src153”. 

Or a copy of your new release??  

By the way, I think I have original floppies for the latest PC MOS  15 user version as well as manuals, and related utility disks etc.

We can run our programs on DOS or XP but really love our PC MOS and would love to keep it alive!
Looking forward to any advise or help you can offer!”

Myron Kraushar
Canada

**Comment by the-grue** [comment]
Hi Myron!  The "new release" is just a build of the code in the repo with a fix for the date bug and some other small fixes by the community.  It is still in release candidate status and I have not gotten any feedback as far as folks testing or even using it.  Until I get some feedback on actual testing, it could possibly stay in RC status forever in my repo.  I don't want to continue to bloat @roelandjansen 's repo with intermediate files.  And my RC isn't the only one out there.  There are several other forks that may include new features.  For example, see @galazwoj 's work.

If you'd like, you can e-mail me your $$MOS.SYS file and I can patch it for you or, if you want, you can see the Linux based patch command submitted by @sub205 in #54 .

Please note that we offer no warranty.  The patch changes a single byte in the file to fix the date bug.

Jim

**Comment by RenownMDK** [comment]
I would really be grateful if you could do that!  I am leaving tomorrow for a week of holiday,
but I will try to get it sent before I go.  

Do you think my issues recently with the CMOS date of 08/07/2007 are related to the same bug?
Odd that we didn’t have any problems when the computer clock reached this date in “real time” 12 years ago.
As I stated we have been running the computer clock 12 year slow (to match leap year and day of week)
and adding 12 years as a variable to our MOS Auto-exec boot up file as a workaround since Aug 1 2012.

Computer also did some wired things this time...all my c:\ boot up files and c:\taskX.bat files were replaced by 
similarly named files deleted years ago after a reboot when the problem manifest a few days ago. Even a couple of task.bat files 
we deleted years ago appeared.

I Managed to get the system up and running using the same dates fudge as above,  08/03/2007 and after seem to read ok.  

It would also be amazing to be able to load price files etc. without having to create blank files with an old date
the overwriting them or loading and saving the files on an old XP computer with an old system date!  

You guys are amazing for doing all this work on a 30 year old operating system!

Myron

**Comment by the-grue** [comment]
I haven't seen the 2007 bug before, then again, I didn't know PC-MOS existed until a couple years ago.

The files reappearing could be a byproduct of the munged date/time entries when things went haywire back in 2012.  The 8/1 matches, but the 2007 doesn't.  If you put a fix in place on 8/1/2012 or 8/2/2012 that could explain why everything works on 8/3.

It would be interesting to see if correcting the date bug makes more files visible.  I'll keep an eye out for your file and try to turn it around quickly.

Jim

________________________________

**Comment by andrewbird** [comment]
@the-grue I still find it pretty amazing that you had the motivation / persistence to track this bug down, well done!
Regarding testing of your RC, i don't have real hardware for PC-MOS so I'm only likely to test on Dosemu. I won't be able to test any of the protected mode stuff as it conflicts with the emulator, or read/write disk access as PC-MOS has no redirector support. I guess what I'm trying to say is I'm loathed to report back anything when coverage would be so little, and I'd rather not give you a false positive indication.

**Comment by the-grue** [comment]
Thanks for the feedback and kind words Andrew!  I am aware of some of the bugs in the protected mode stuff.  It works fine on VirtualBox and one laptop I have tested it on but it fails on another laptop and a desktop I have tested it on.  Any real fix to the PM drivers would have to wait until the next version.  My idea for the RC is to get what we have in an installable distribution with the latest fixes provided by the community.  After that, it would be good to merge in some of the fixes/changes out there on other forks.

Any testing has value, so if you come across anything acting strangely in the standard real mode DOS stuff, please let me know.  I am currently working on a FAT32 driver for a different project but monitor this project daily to see if there is any activity or things I can chime in on.

If anyone is really holding off playing with MOS waiting for the RC to go GA I can revisit it, pretty it up, and create a pull request.

Jim

________________________________

**Comment by stsp** [comment]
> if you come across anything acting strangely in the standard real mode DOS stuff, please let me know.

Seriously? :)
Will you take a 128+ bug reports if someone
will try his old games collection or something
else real-mode? The lack of fs redirector makes
such testing a bit problematic, but preparing
the hd image is not a big problem after all. I just
don't suppose you want some real testing to
happen, or am I wrong?

**Comment by RenownMDK** [comment]
Attached $$MOS.SYS file, hope it gets thru as is.
I checked the original install disks, its is a 33 user PCMos, Release 5.01  UDP6  

Myron

**Comment by the-grue** [comment]
Hi @stsp !  Within reason, I would imagine ANY issues could be added to this forum to be worked through if someone wants to take the issue on.  I am just talking about some basic testing so we can get the new build out of RC status.  OS issues, I'll take a look at.  Application compatibility issues, not so much since I wouldn't be able to test them anyway without having the applications myself.

As far as I know, I built the RC and have used it quite a bit myself, I haven't gotten feedback from anyone else that is using it.  If anyone would be open to me throwing the RC into GA and letting anyone test it in the field I am ok with that.  As long as the community will help support and resolve issues.  

Honestly, if it worked on 5.01, it should work on my 5.02 rc.

Jim

**Comment by the-grue** [comment]
No attachment.  Please send it to me directly.  My handle at hotmail.com.

Jim

________________________________

**Comment by stsp** [comment]
> Application compatibility issues, not so much since I wouldn't be able to test them
> anyway without having the applications myself.

If you want this to happen, I can arrange that:
create and provide some hdimages with various
apps. But I don't think we are at the level when you
can't go to some abandonware site and try a few
progs or games on your own - I am pretty sure you'll
trigger the bugs quickly even w/o my help.
By the way, you said the PM drivers (of mos) should
not be tested as known buggy, but should the
unrelated PM apps be tested?

> Honestly, if it worked on 5.01, it should work on my 5.02 rc.

OK, I see.

**Comment by the-grue** [comment]
> > Application compatibility issues, not so much since I wouldn't be able to test them
> > anyway without having the applications myself.
> 
> If you want this to happen, I can arrange that:
> create and provide some hdimages with various
> apps. But I don't think we are at the level when you
> can't go to some abandonware site and try a few
> progs or games on your own - I am pretty sure you'll
> trigger the bugs quickly even w/o my help.

You'll understand what I mean if you think about it this way.  I would have to go download an application I am unfamiliar with, install it, then try to figure out how to use it and determine why it is failing.  I have tested with stuff I am familiar with and have had good results.  Mostly development tools at this point.  I haven't tried any of my old DOS games.  As an example, Borland C++ 3.1 works fine.  I was able to get GEM to work.

Really, all I would like to see is someone or a couple/few people just download the rc and play around with it to see if it works the same as 5.01 and let me know their findings.  If it does, great, I'll do a pull request and get it out there.  If something behaves terribly wrong, then it will need some attention before kicking it over.

Or maybe I am putting too much thought into this and should just consider the 5+ months it has been available in rc good enough.

Jim

**Comment by the-grue** [comment]
> But I don't think we are at the level when you
> can't go to some abandonware site and try a few
> progs or games on your own...

I would like to take this opportunity to point out that (I hope) I am not the only person working on this project and comments like this are really not very productive.  I don't think asking for some folks to invest a little of their time to do a little testing is too much of a burden.

Help/don't help.  Contribute/don't contribute.  But don't be calling me lazy.

Jim

**Comment by stsp** [comment]
> As an example, Borland C++ 3.1 works fine.

Wow! Protected-mode IDE that compiles and runs
things - great result!

> But don't be calling me lazy.

That's not what I meant, just assumed that downloading
the pre-built hdimages or games from sites is nearly the
same thing wrt spent efforts. You'll have to do the d/l+run
cycles in both cases. But if you prefer the hdimage way, and
especially given that things like bc ide works fine,
some testing is really a due.

**Comment by andrewbird** [comment]
It's a funny old thing this project, for months nothing, then a flurry of activity, then back to just a trickle of posts. I can't say why, but it certainly reflects my contribution pattern too.

**Comment by sub205** [comment]
A good way for testing is PCEM. It's based on qemu, but highly sophisticated. Can emulate many PCs including BIOS.

You can find it here: https://pcem-emulator.co.uk/
There are also the corresponding ROM-collections out there (google for it) - they are not included for legal reasons.

**Comment by stsp** [comment]
@sub205 dosemu works perfectly for both building
and running mos. If mos had an fs redirector, there
would also be a host fs access, but its not.

**Comment by the-grue** [comment]
See pull request #82

**Comment by sub205** [comment]
@stsp But afaik dosemu can only run the user applications, not the kernel itself, which leaves a big part of the operation system aspect out of scope.

**Comment by stsp** [comment]
This is true only for protected-mode
kernels which it doesnt run. For real
mode it runs both app and a kernel,
and as long as we do not test the mos
protected-mode drivers, it can be used
fully. When the prot-mode drivers are
in a testible state, likely dosemu will be
out of any help indeed. (as you say,
it only run protected-mode apps, not a
ring0 stuff)

**Comment by stsp** [comment]
You probably mean dosbox which
usually run only dos apps (but still
have an ability to boot an alien dos
kernel if you try hard enough).
dosemu is designed to run real-mode
dos kernels.

**Comment by sub205** [comment]
Ah damn it, of course you're right. To early, to less coffee. :-)
Anyway, give pcemu a shot, it's very comprehensive. The emulation is so good, you can even run stuff like copper bars (text mode effect, where you count scanlines and in retrace reprogramm the palette dac for illusion of graphic) and other demo-scene-stuff on it. Also, you get different CPUs and replicas from many x86-plattforms (xt, tandy, 286 etc up to pentium)

**Comment by stsp** [comment]
Or maybe you mean older dosemu,
i.e. I think @andrewbird wrote a loader
for mos kernel just a few years ago,
so on some dosemu builds this is indeed
unsupported.

**Comment by stsp** [comment]
Is there a need to ever try pcem over
dosbox? dosbox should run all the
effects too. pcem may likely be
adequate only for mos, as mos have
no fs redirector built in, so other dos
emulators can give no advantage of
it to give you a host fs access.
For other doses I really am yet to find
out any advantage of pcem over
dosemu or dosbox. :)

**Comment by stsp** [comment]
By the way, I think also MESS provides
a very accurate emulation of early PCs,
but I never have tried that as well.
Always thought that a suited dos
emulators run dos better than the
generic ones. :)

**Comment by sub205** [comment]
Because you have an entire system that works (nearly) exactly as the original hardware, including protected mode.

Let us sum it up:

- Dosbox has it's own interpreter and is more like "wine", providing emulated systemcalls.
- Dosemu seems to me like a stripped down version of qemu with some addons for better dos usability.
- QEMU is a real hardware level emulator providing many different cpus and options.
- PCEMU is a QEMU-derivat, optimized to even simulate different chipsets, voodoo graphics, CPU-speeds etc to get a nearly 100% emulation.

Every system has it's pros and cons and a good place to use it. It's up to the user to chose the best tool for the job. ;-)

**Comment by stsp** [comment]
qemu-derived code-wise you mean?
I haven't heard it reuses any qemu
code, in fact I thought pcem has a cpu
interpreter whereas both dosbox and
qemu have JITs (and of course qemu
just uses vtx instead of jit).
Have you also tried mess btw? I think
it is very comparable to pcem by design,
but someone have to look at actual
results.

**Comment by sub205** [comment]
Hm, you're right. No references to qemu, despite its similarities in many technical aspects.
They have a jit, but it's not that optimized.
I'll give mess a shot!

**Comment by stsp** [comment]
This level of similarity is common to
all the above emulators: even dosbox,
besides just implementing the dos
calls as you mentioned, also emulates
chipset, voodoo gfx etc. But at that
point similarities end, as for example
only pcem and mess go as far as to
require the native firmware files for
different components. qemu does
not emulate on that level (it is mostly
concerned by a speed).
Something that AFAIK only dosemu
provides, is what you call an addons
for better dos usability. I dont know
how you boot mos under pcem, but
I suppose you use floppy image, which
you would need to update after changing
the mos code. dosemu has addon
(loader) to just boot mos from anywhere,
for example from built source tree,
which avoids the image rebuilding step.

**Comment by stsp** [comment]
Oh wait, it seems (contrary to what I
said before) that dosemu does even
provide some host fs access to mos,
but it seems read-only.
I'll fill a bug report.

**Comment by andrewbird** [comment]
> dosemu has addon
(loader) to just boot mos from anywhere,
for example from built source tree,
which avoids the image rebuilding step.

Interesting, sounds like grounds for travis-ci build and test.

**Comment by andrewbird** [comment]
> dosemu does even
provide some host fs access to mos,
but it seems read-only.
I'll fill a bug report.

Yes will help for other DOS that have an incompatible or missing redirector, for example early DR-DOS (3.3x to 6.0)

**Comment by stsp** [comment]
> Yes will help for other DOS that have an incompatible or missing redirector

Filled in the bug report.

As for pcem, I am really surprised with how many
devices does it support:
https://pcem-emulator.co.uk/status.html
and yet I don't fully understand its mission.
In my eyes it:
- will never be as fast as qemu
- will never provide any compatibility advantages
for DOS software over dosbox
- will never provide any usability advantages for
realmode OSes over dosemu
- will have to compete with other generic emulators
like bochs or mess for the niche that pretty much non-existent
(the niche of mess is mostly in an emulation of game
consoles, it emulates PC just "by the way")

Of course the above assumptions may be entirely wrong.

**Comment by sub205** [comment]
In compatibility, it's far better than dosbox ever can be because it tries to 100% emulate the behavior of the real hardware. Dosbox only emulates the dos-api and some lowlevel-stuff.

Performance is no argument, because you only need to run these machines at 100% of their native speeds. There is simply no need to emulate a 386 and have it perform like a i5. You can run the corresponding software on real hardware because it's still readily available. And the software that benefits from emulation has no need for faster CPUs than from the corresponding periode.

Yes, everything is a niche, even this project. But there are users and developers that continue to work on/with it ;-)

Personally, i love PCem because i can run/build native images for period-correct PCs, copy them over to CF-card and run them on real hardware if i want to. I'm a retro enthusiast. And with PCem you can kinda virtualize your hobby before you need to build real hardware.

**Comment by stsp** [comment]
> In compatibility, it's far better than dosbox ever can be because it tries to 100%
> emulate the behavior of the real hardware. Dosbox _only_ emulates the dos-api
> and some lowlevel-stuff.

I think this is an underestimation of dosbox, in particular
the word "only" is a culprit. While I never used pcem, I used
dosbox a lot, and it is designed to be 100% accurate
on emulating everything needed by DOS software (games).
I don't think it provides the CPU cycle-perfect timing, but
for all timing-sensitive tasks like demos, it works very well.
So do you know the real examples where dosbox fails and
pcem succeeds? I think if you do, these progs will work
correctly only on a particular CPU for which they were timed,
which I think is quite rare (and of course dosbox can be
manually slowed down for pathological cases).

> And the software that benefits from emulation has no need for faster CPUs
> than from the corresponding periode.

This is true, but the things are more complex.
When you emulate voodoo card and do an fm or midi
synth, and run the CPU interpreter in parallel, it appears that
you can't run some "heavy" DOS game with an adequate
speed. Even dosbox suffers from that _a lot_, despite the
good jit. So I am pretty sure this gonna be a very big problem
for pcem - if not, that wouldn't be such a big problem for
dosbox but it is.
So, in short, you can emulate the CPUs from old era with
their real speed, but when it comes to emulating the whole
PC, things are getting much more complicated and keeping
up the CPU performance became a real challenge.

> Personally, i love PCem because i can run/build native images for period-correct PCs

Yep, testing the image on different emulated PCs
to see where it suits, and then put it exactly on that
PC, looks like a good use-case, thanks, now I know
the niche exists also for pcem. :) I do not suppose
MESS emulates so many old PC clones, so this feature
is probably indeed unique to pcem.

**Comment by stsp** [comment]
Mm, I pulled the hg repo of pcem, and the last
commit seems to be dated Sun May 19 17:41:11 2019 +0100 -
3 months ago.
@sub205 did they move the repository elsewhere,
or the project is just very slow at development?

**Comment by roelandjansen** [comment]
afaik this is the place for the repo indeed and it seems a bit slow. 

On the other hand, nobody probably gets paid so ... ;-)

**Comment by sub205** [comment]
indeed, i had plans to do an automated buildscript for linux, unfortunately had many projects inbetween and got a bit sidetracked. next project: my son, estimated arrival: 4-5 weeks. I hope to find some spare time to do work here. :-)

**Comment by stsp** [comment]
@roelandjansen are you talking about pcem?
So you also use pcem to run mos? :)

**Comment by stsp** [comment]
> or the project is just very slow at development?

Just to clarify, the question was about the pcem repo.

**Comment by the-grue** [comment]
I use DosBox quite a bit for my projects that require 16-bit tools.  I believe the only real limitation it has is that it doesn't support above the 386 instruction set (and that may be old information).  Ok, a second one would be that the cli doesn't support full/reliable functionality compared to a DOS, but that's just being picky.  

From there I use VirtualBox for further development and testing before moving on to real hardware.

I have used a lot of emulators/simulators over the years and these ones work for me for my current projects.  Unless I go back to hacking PDP/11 or VAX code, then it is simh all the way!

**Comment by stsp** [comment]
> As an example, Borland C++ 3.1 works fine.

For me bc just hangs tight on start.
The testing is troubled though, as the patches
I provided, seems not included in 5.02,
SHIPMOS archive is not updated with 5.02
(the best I could find was a floppy image)
and the build procedure seems problematic
again (need to investigate why, it used to
build for me in the past).
I managed to cook something bootable,
but bc hangs, but may be a bad build...
Will looks further.

**Comment by andrewbird** [comment]
> and the build procedure seems problematic
again (need to investigate why, it used to
build for me in the past).

The MAKEMOS.bat file builds in different directories to those @the-grue put the latest stuff in. His build/install was done using the MAKEALL.BAT but is unfriendly to the dosemu & build.sh script, so you probably need to tweak one or the other.

**Comment by stsp** [comment]
OK, I managed to get the build.sh going
after some fsck stuff, but the fact that the
boot patches were not applied, is disappointing.
And I am not sure where are the release files,
as SHIPMOS.ZIP is still old.

**Comment by andrewbird** [comment]
@stsp I think the patch was applied in kernel directory, but since there are too many almost identical directories, it's not surprising that it was omitted. Only a good tidy up will fix that, and no more file or directory copies going forward.

**Comment by stsp** [comment]
`speed 0` does indeed solve all your speed complains.

**Comment by stsp** [comment]
(oops, wrong thread... but nevermind... it was about building MOS)

**Comment by andrewbird** [comment]
Not for me, I'm afraid, but I'll worry about speed later.

**Comment by the-grue** [comment]
All of the latest files for the kernel, drivers, and utilities are in "latest".  I made no assumption that anyone was using any particular emulator/operating system/etc. just MOS to do the builds, so my build scripts work with an installation of MOS.  I believe it will also work with DOS, but haven't built under DOS in awhile so I don't remember.  As for it being unfriendly to dosemu, it should work if dosemu is actually running MOS.  :)

If you are looking for files, they are all in the IMAGES/v502/MOS502.img file.  It is modelled after the actual distribution provided by TSL, so if you mount that floppy image, you will see all the files.  Granted, some of them are in the compressed .exe files but that is the way TSL did it, so I just followed suit to keep things familiar.  If you do an install, it will extract everything in a familiar fashion.

If you run the makeall.bat script, it should build all of the files into /DISTRO.  If you stop it short of the full processing of the build with either a pause in the .bat file or <ctrl> c you should be able to see all of the files before they are packed and copied to a floppy.

@stsp, you mention some patches you provided.  When were they provided and where did you put them?  If they were in before I did my code audit, they should be included in the files in the "latest" directory or other directory if you made changes to, for example, acu in the acu directory.

**Comment by the-grue** [comment]
> For me bc just hangs tight on start.

Do you have the protected mode driver installed?  There are conflicts if you do.  Try it in real-mode.

**Comment by stsp** [comment]
> just MOS to do the builds, so my build scripts work with an installation of MOS.

I was talking about my cross-compile build.sh.
Haven't tried building MOS from MOS.

> If you are looking for files, they are all in the IMAGES/v502/MOS502.img file.

Yes, this is the only place I've found, so I
extracted the files from there (with loop mount)
and found they lack my patches. I hope this is
still a pre-release so the things can be corrected.
Given that SHIPMOS.ZIP was not yet updated,
I guess so.

> When were they provided and where did you put

https://github.com/roelandjansen/pcmos386v501/commit/37cf1793606ba5f320faf0bbead0d04a2536ed37
https://github.com/roelandjansen/pcmos386v501/commit/5fa7b36d9dbe10268665cdb3ba8f325505925147

AFAICT neither one is included in the files I extracted
from an image, as I am having the same boot troubles
as before.

**Comment by stsp** [comment]
> Do you have the protected mode driver installed? There are conflicts if you do. Try it in real-mode.

I used the default config.sys:
```
DEVICE=\CDROM\ATAPI_CD.SYS /D:MSCD000
DEVICE=\CDROM\CDLINK.SYS D=MSCD000
smpsize=64
cache=16
LASTDRIVE=Z
```

and default autoexec:
```
batecho off
envsize 4096
rem install
```
(rem is mine)
and after that tried the (protected-mode) bc.exe

**Comment by andrewbird** [comment]
@the-grue sorry if my comment seemed overly strong, it wasn't my intention to be harsh or critical. I was trying to get across that the new MAKEALL.BAT script has some stuff in there like drive references and paths (to floppy etc) that won't necessarily exist on a default dosemu boot and and that some mods would be necessary. Keep doing the good work!

**Comment by roelandjansen** [comment]
that reminds me of that I used to build under mos itself. 

What do we think to be the best building env? 

I  played a bit with the initial files and could not really have a large enough disk to be able to build.
Maybe I'll redo now with the latest code.

**Comment by the-grue** [comment]
> @the-grue sorry if my comment seemed overly strong, it wasn't my intention to be harsh or critical. I was trying to get across that the new MAKEALL.BAT script has some stuff in there like drive references and paths (to floppy etc) that won't necessarily exist on a default dosemu boot and that some mods would be necessary. Keep doing the good work!

No worries @andrewbird .  I was just trying to highlight what my build/execution environment looks like and why I felt it was important for others to do some testing while the new build was in RC status.  But this is fine too.  At least it is getting some testing and giving us some things to work on.

I just can't test everything or everyone's environments for them.  :)

**Comment by the-grue** [comment]
@stsp how did you install bc?

**Comment by stsp** [comment]
> I would try removing the CD-ROM drivers, reduce smpsize to 16 and try again.

Hmm, good shot (almost): the problem
disappears after removing `cache=16`

**Comment by the-grue** [comment]
> > When were they provided and where did you put
> 
> [37cf179](https://github.com/roelandjansen/pcmos386v501/commit/37cf1793606ba5f320faf0bbead0d04a2536ed37)
> [5fa7b36](https://github.com/roelandjansen/pcmos386v501/commit/5fa7b36d9dbe10268665cdb3ba8f325505925147)
> 
> AFAICT neither one is included in the files I extracted
> from an image, as I am having the same boot troubles
> as before.

A quick check proves you are correct.  The updates you applied were to the kernel directory.  As a part of my code audit and the creation of the "latest" directory, which includes all of the latest changes previously in UPDAT501.SYS, these files were most likely copied from the "mos5src" directory.

This was part of the problem when trying to identify and collect the latest files.  Folks were making changes in multiple directories, which made it hard(er).

Would it be possible for you to apply your updates to the files in "latest"?  Or I can add them in when I get some time if you would rather not.

**Comment by the-grue** [comment]
> > I would try removing the CD-ROM drivers, reduce smpsize to 16 and try again.
> 
> Hmm, good shot (almost): the problem
> disappears after removing `cache=16`

Good thing I am not a gambler!

**Comment by the-grue** [comment]
@stsp I reapplied your fixes and am going to rebuild/redeploy.  Should have a floppy image to test with shortly.

**Comment by stsp** [comment]
Do you have all keys in bc working properly?
I have only a few keys working, can't navigate
the menus etc.

> I reapplied your fixes and am going to rebuild/redeploy. Should have a floppy
> image to test with shortly.

I tried, but getting "Improper MOS version" error.
When built from `kernel` - works fine.

**Comment by the-grue** [comment]
Yeah, I am able to navigate the menus no problem.  I haven't tried all key combinations I am sure, but my menus are working fine.

Not sure about that error.  Building from kernel will build version 5.01.  You would need to change the minor version in file VERSION.INC from 1 to 2.

**Comment by stsp** [comment]
> Yeah, I am able to navigate the menus no problem.

It appears I can navigate with the keypad keys -
only arrow keys do not work. Please confirm that
you can navigate with an arrow keys.
Other than that, compiled a small protected-mode
example, and it works! Amazing. So some serious
tests can be done indeed.

**Comment by the-grue** [comment]
Ok, give this a try.

[STASTEST.img.gz](https://github.com/roelandjansen/pcmos386v501/files/3489309/STASTEST.img.gz)

**Comment by the-grue** [comment]
> Please confirm that
> you can navigate with an arrow keys.

Confirmed I can navigate with the arrow keys as well as the keypad.

**Comment by stsp** [comment]
> minor version in file VERSION.INC from 1 to 2.

Yes, that was the problem.
Fixed here:
https://github.com/roelandjansen/pcmos386v501/pull/86

**Comment by stsp** [comment]
> Ok, give this a try.

This image doesn't include bc (it won't fit anyway).
But your confirmation is enough.
I'll see if there is something on my side.
On the other front:
```
[C:\]bin\mem
 
Memory Type        Total       Used       Free
----------------  --------   --------   --------
Conventional          640K       250K       390K
Upper                   0K         0K         0K
Reserved              384K       384K         0K
Extended (XMS)      8,256K     8,256K         0K
----------------  --------   --------   --------
Total memory        9,280K     8,890K       390K
 
Total under 1 MB      640K       250K       390K
 
Largest executable program size       390K (399,184 bytes)
```

No deal. Most progs fail to start.
Any cure to get more memory?

**Comment by the-grue** [comment]
> This image doesn't include bc

That is a rebuild of the v5.02 installation disk with your patches re-introduced.  I wanted you to try it and see if it works as expected now.

As for the memory, I don't really know what I am looking at.  That looks like DOS.  If you want to show what MOS says, use mos map and mos info.

**Comment by stsp** [comment]
> Confirmed I can navigate with the arrow keys as well as the keypad.

Dosemu have various tweaks (directly in the
code unfortunately, as defines) to slow down
keyboard. I enabled one, and that helped.
Now all keys work, wonderful.

Will you be able to fix the 5.02 with the boot
patches?

**Comment by stsp** [comment]
> That is a rebuild of the v5.02 installation disk with your patches re-introduced.
> I wanted you to try it and see if it works as expected now.

Ah, yes, indeed.
Tried - no boot bug.
All works as expected.

**Comment by the-grue** [comment]
Cool!  I'll push updates.

**Comment by stsp** [comment]
> As for the memory, I don't really know what I am looking at. That looks like DOS.

mem from freedos. Why not?
Anyway:
```

[C:\]mos info
 
PC-MOS System Information         Start     End
 
Freemem =                         NO <NOT RELOCATED>
 
MOS Kernel Segment #1               700     D200
MOS Kernel Segment #2              D200    1B7F0
System Memory Pool (SMP) #1       1B7F0    2B7F0
Disk Cache Descriptors            2BA00    2BA80
Disk Cache                        2BA80    2FA80
Command Processor                 2FA80    37E90
 
 
[C:\]mos map
 
                         PC-MOS USER TASK STATISTICS
 
Task Start Size Video  User    Program     Port   Baud   Pri Slice Files  Status
 
 0*  37E90 416K  CGA         MOS     .COM   N/A    N/A   2    1     3   ACT  ND
 
   14K of    64K SMP Allocated
```

Does it tell you something?

**Comment by the-grue** [comment]
Just confirms what the mem command did.

Try reducing your smp from 64K to 16K, since you are only using 14K.  Also, don't load the CD-ROM drivers if you aren't using it.  I have my smp set to 4 or 8 when not loading drivers and set the cache to 2 because the virtual machine should be fast enough with/without the cache.

Other than that, you would need to load the 386.sys, which would load a lot of the stuff high and free up about 500K or so.  You can get it ~ 600 if you fiddle with it enough.

I don't know if you can load anything high without enabling protected mode.  Then you have a whole new set of problems.

**Comment by stsp** [comment]
> I don't know if you can load anything high without enabling protected mode.

dosemu provides drivers exactly for that,
and surprisingly they work with MOS:
```
dosemu XMS 3.0 & UMB support enabled
dosemu EMS driver rev 0.8 installed.
5K SMP #1 allocated;     2K of SMP #1 allocated to device drivers.
 
[C:\]batecho off
 
[C:\]bin\mem
 
Memory Type        Total       Used       Free
----------------  --------   --------   --------
Conventional          640K       202K       438K
Upper                   0K         0K         0K
Reserved              384K       384K         0K
Extended (XMS)      8,256K        64K     8,192K
----------------  --------   --------   --------
Total memory        9,280K       650K     8,630K
 
Total under 1 MB      640K       202K       438K
 
Total Expanded (EMS)                8,576K (8,781,824 bytes)
Free Expanded (EMS)                 8,128K (8,323,072 bytes)
 
Largest executable program size       438K (448,336 bytes)
```

As you can see, UMB and EMS drivers both
reported success, and EMS is confirmed by
"mem". But UMB requires `DOS=UMB` or
`DOS=HIGH,UMB` in config.sys, which MOS
doesn't understand. Is there an equivalent?

**Comment by the-grue** [comment]
What does mos info and mos map say now?

**Comment by stsp** [comment]
```
[C:\]mos info
 
PC-MOS System Information         Start     End
 
Freemem =                         NO <NOT RELOCATED>
 
MOS Kernel Segment #1               700     D200
MOS Kernel Segment #2              D200    1B7F0
System Memory Pool (SMP) #1       1B7F0    1F7F0
Disk Cache Descriptors            1FA00    1FA80
Disk Cache                        1FA80    23A80
Command Processor                 23A80    2BE90
 
 
[C:\]mos map
 
                         PC-MOS USER TASK STATISTICS
 
Task Start Size Video  User    Program     Port   Baud   Pri Slice Files  Status
 
 0*  2BE90 464K  CGA         MOS     .COM   N/A    N/A   2    1     3   ACT  ND
 
    4K of    16K SMP Allocated
```

Grr, not much different, still as cryptic. :)

**Comment by the-grue** [comment]
What's your config.sys look like?  I have only seen it push stuff high when you use 386.sys.

**Comment by stsp** [comment]
```
;DEVICE=\CDROM\ATAPI_CD.SYS /D:MSCD000
;DEVICE=\CDROM\CDLINK.SYS D=MSCD000
DEVICE=\DOSEMU\UMB.SYS
DEVICE=\DOSEMU\EMS.SYS
DOS=UMB
smpsize=16
;cache=16
LASTDRIVE=Z
```

`DOS=UMB` here produces an error.
On all other DOSes it would activate an UMBs
(and in real mode, since the drivers from dosemu
are real-mode ones).

**Comment by the-grue** [comment]
I uploaded the v5 documentation to the repo and it is checked in.  Might look in there under configuration.  MOS might not have a specific line like that native to it?

**Comment by stsp** [comment]
So it really insists on a native driver for
UMB rather than to use the generic APIs
for that? That would be a severe limitation...

Also tried dosemu\emufs.sys (rw host fs access)
and it says `EMUFS revectoring only` which
is some kind of an error, but R/O access is
still there and that's very good.
Now to the docs...

**Comment by the-grue** [comment]
Not saying it needs a native driver.  Just that DOS=UMB may not be the proper syntax, if there is even such a thing in MOS.

Oh, please see #88.  Think I got it right this time.

**Comment by stsp** [comment]
Btw, I would suggest to disable that `cache=`
stuff as it causes troubles, or if you want, I
can try other values of that directive.

**Comment by stsp** [comment]
Yes, #88 looks right.

**Comment by stsp** [comment]
> Btw, I would suggest to disable that cache=

Oh wait, no, I wouldn't!
With your files I can't reproduce the problem
with cache any more. So it was because of the
kernel/mos5sources build got messed up likely.

**Comment by stsp** [comment]
So my config.sys is now:
```
;DEVICE=\CDROM\ATAPI_CD.SYS /D:MSCD000^M
;DEVICE=\CDROM\CDLINK.SYS D=MSCD000^M
DEVICE=\DOSEMU\UMB.SYS^M
DEVICE=\DOSEMU\EMUFS.SYS^M
DEVICE=\DOSEMU\EMS.SYS^M
FREEMEM=B0000,B8000^M
FREEMEM=D0000,E0000^M
smpsize=16^M
cache=16^M
LASTDRIVE=Z^M
```

And mos info:
```
[C:\]mos info
 
PC-MOS System Information         Start     End
 
Freemem =                         B0000    B8000
Freemem =                         D0000    E0000
 
MOS Kernel Segment #1               700     D200
MOS Kernel Segment #2              D200    1B7F0
System Memory Pool (SMP) #1       1B7F0    1F7F0
Disk Cache Descriptors            1FA00    1FA80
Disk Cache                        1FA80    23A80
Command Processor                 23A80    2BE90
```

So we are on a right track, but:
```
[C:\]bin\mem
 
Memory Type        Total       Used       Free
----------------  --------   --------   --------
Conventional          640K       202K       438K
Upper                   0K         0K         0K
Reserved              384K       384K         0K
Extended (XMS)      8,256K        64K     8,192K
----------------  --------   --------   --------
Total memory        9,280K       650K     8,630K
 
Total under 1 MB      640K       202K       438K
 
Total Expanded (EMS)                8,576K (8,781,824 bytes)
Free Expanded (EMS)                 8,128K (8,323,072 bytes)
 
Largest executable program size       438K (448,336 bytes)
[C:\]mos map
 
                         PC-MOS USER TASK STATISTICS
 
Task Start Size Video  User    Program     Port   Baud   Pri Slice Files  Status
 
 0*  2BE90 464K  CGA         MOS     .COM   N/A    N/A   2    1     3   ACT  ND
 
    4K of    16K SMP Allocated
```

So even though the `FREEMEM=` commands
worked, mos didn't relocate itself and didn't free
any low mem.

**Comment by the-grue** [comment]
Take a look at MEMDEV on page 2-24.

Where are the UMB.SYS, EMS.SYS, etc. coming from?

When I set MEMDEV=$386.SYS /P and reboot it loads kernel segment 1 to FREEMEM, kernel segment 2 to UMB, and everything else but the command processor and master video context area to FREEMEM and I have 552K size shown in mos map.

Of course, this enables the MOS protected mode stuff and if you have it loaded, Borland C doesn't like it.  It will throw an exception if you try it.  If you have EMM loaded, it will complain:

DPMI server initialization error->v86 task without vcpi.

Sorry, didn't want to add another screenshot.

**Comment by stsp** [comment]
Do you think its right to not handle
FREEMEM in real mode? It doesnt
seem to be documented to require
any MEMDEV driver.
Also 552K is still a no go for most
games. 620-630 is normal.

**Comment by the-grue** [comment]
I don't believe MOS was designed with games in mind.  I get memory issues when trying to install Windows and several applications.  GEM is the only GUI I have had any luck with.

Now, if we want to strip out all of the multi-user stuff and create a true DOS clone, we could possibly get closer to that goal.  It would probably save alot of memory to remove the security stuff, etc.

As to whether I think it is right or not, if you keep assuming MOS == DOS then you will be running up against these issues at every turn.  Do I think it is right for MOS, yes, because I understand it is a different OS.  Do I think it is right for DOS compatibility, no, because of what you have pointed out.

**Comment by roelandjansen** [comment]
It certainly was not designed to be a games platform, on the contrary. 

It tried to compete with multitasking/multiuser examples like desqview but then without the issues like memory constraints etc.

But as @the-grue states -- MOS != DOS. If you want to have compatibility improvements so that 
all kinds of older s/w still runs (like a database with a front-end and some other things, also maybe over serial lines) it's fine. Making it a dumb clone... I'd say -- we have already  a lot of them who do a quite good job and are open as well.

When it comes to memory issues, I still believe stuff like DPMI could help a bit there but I also understand that it's a headache-thing.

**Comment by roelandjansen** [comment]
I looked at some of the asm code that hamdles the memdev and freemem things and could not directly conclude that memdev is required or not. Looking at how mos goes when 386.sys is loaded -- it shows that a part is aboven the 1 MB mark already. I used a 5 user version on a 8M machine and it outperformed all I had at that time. All the tasks were sized well enough. I think that finetuning a bit might get me around the 600k mark. But then, its rnning in protected mode; one of the major plusses it has plevening blowing up other tasks.

**Comment by stsp** [comment]
MOS was not designed for games, but
games do represent a good test-cases
for DOS compatibility. Well, other than
games, perhaps another good test-case
would be a 32bit borland tool-set etc, but
still the major test-suit I have, consists of
the games.

I do not suggest to remove any multitasking
features. In fact I only suggested to allow
UMBs w/o any prot-mode driver.
But if I am up to suggest something, I'd
say that the things should be configurable.
And ideally, if the one didn't set up any
multitasking switches, he'd better give the
sense of the "dumb DOS clone". But as
soon as he configured in some interesting
driver, the things gets different.
